### PR TITLE
fix: harden agent execution reliability

### PR DESCRIPTION
**Pull request by qinxuye**
## Summary

- recover malformed and unavailable provider tool calls in Auto and ReAct with one bounded, schema-aware retry
- safely repair structurally complete DeepSeek JSON arguments while preserving bounded original diagnostics for tracing
- forward child Auto and DAG-step LLM errors with the correct execution metadata
- make DAG replanning/completion checks respect valid dependencies and the user's authoritative requested scope
- require trusted tool-result provenance before claiming delivery of a generated file
- make checkpoint blob persistence idempotent under concurrent DAG writes
- report the image model that actually served generation and edit requests

## Why

Provider protocol failures and parallel execution races were terminating otherwise recoverable tasks or hiding the original failure details. In other cases, narrowed tool schemas, invalid replans, or unverified file references could produce a failed run or an incorrect completion claim.

These are general agent-runtime reliability fixes and do not contain the static visual design skill, ad-creative art direction, logo handling, or brand-asset workflow.

## Relationship to #942

This PR extracts the non-ad-creative reliability work from #942. The #942 branch has been left unchanged and will be rebased onto main after this PR merges.

## Validation

- 332 focused tests passed across Auto, DAG, ReAct, DeepSeek/OpenRouter protocol handling, image model reporting, checkpoint persistence, and trace storage
- commit hooks passed, including Ruff, formatting, mypy, isort, and codespell
- git diff --check upstream/main...HEAD

### CODE REVIEW

File: src/xagent/core/model/chat/basic/deepseek_tool_protocol.py [hunk 1]
```diff
@@ -262,10 +326,106 @@ def _tool_call_violation(
...
+
+        if char == '"':
+            in_string = True
+        elif char in "{[":
+            stack.append(char)
+        elif char in pairs:
+            if not stack or stack.pop() != pairs[char]:
+                return False
+
+    return not in_string and not escaped and not stack
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/core/model/chat/basic/deepseek_tool_protocol.py]: Fixed in 1ec05b6f. The completeness scanner now tracks the active quote character for both single- and double-quoted strings, including escapes. Added a regression case with an unmatched brace inside a single-quoted value.

File: src/xagent/core/model/chat/basic/deepseek_tool_protocol.py [hunk 2]
```diff
@@ -215,27 +240,65 @@ def _tool_call_violation(
...
+                applied = _set_tool_call_arguments(
+                    tool_call,
+                    json.dumps(arguments, ensure_ascii=False, separators=(",", ":")),
+                )
+                if not applied:
+                    argument_details["repair_status"] = "repair_application_failed"
+                    return _malformed_arguments_violation(
+                        name,
+                        details=argument_details,
+                    )
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/core/model/chat/basic/deepseek_tool_protocol.py]: Fixed the misleading non-dict repair status in 1ec05b6f: repaired lists/primitives now fail immediately with repair_status=failed_non_dict. I intentionally kept an empty dict valid because {} is legitimate arguments for a zero-parameter tool; regression tests now cover both the non-dict failure and empty-object success cases.

File: src/xagent/core/agent/pattern/react/react.py [hunk 3]
```diff
@@ -530,11 +612,15 @@ async def _run_tool_calling_loop(
...
+                    force_final_answer_now = False
                 self.last_response = response
                 normalized = self._normalize_llm_response(response)
                 if self._response_requires_tool_protocol_retry(
                     normalized,
                     force_final_answer=force_final_answer_now,
+                    reject_mixed_control_calls=recover_full_tool_set,
                 ):
                     if answer_streamer is not None:
                         await answer_streamer.fail("invalid tool protocol after retry")
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/core/agent/pattern/react/react.py]: No change needed: the current block already checkpoints at lines 627-632 and immediately returns PatternResult(success=False, status=invalid_tool_protocol) at lines 633-643. The existing test_react_closes_invalid_retry_final_answer_stream_before_failure exercises this exact second-validation-failure path and confirms execution stops; it passes on the current head.

File: src/xagent/core/model/chat/basic/deepseek_tool_protocol.py [hunk 4]
```diff
@@ -262,10 +332,106 @@ def _tool_call_violation(
...
+            return True
+        tool_call["arguments"] = arguments
+        return True
+
+    function = getattr(tool_call, "function", None)
+    try:
+        setattr(function, "arguments", arguments)
+    except (AttributeError, TypeError, ValueError):
+        return False
+    return True
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply src/xagent/core/model/chat/basic/deepseek_tool_protocol.py:413]: I am keeping the current failure behavior. For object-shaped calls, _function_payload treats tool_call.function as the authoritative payload. If writing function.arguments fails, setting a new outer arguments/args attribute and returning success would leave the authoritative function arguments malformed while allowing normalization to proceed. When function is None, validation rejects the missing function name before this setter is reached. Failing repair is safer than reporting a write that downstream parsing will not observe.

File: src/xagent/core/agent/pattern/react/react.py [hunk 5]
```diff
@@ -494,21 +543,49 @@ async def _run_tool_calling_loop(
...
+                    response=response,
+                    metadata=end_metadata,
+                )
             if requires_protocol_retry:
+                if protocol_retry_performed:
+                    if answer_streamer is not None:
+                        await answer_streamer.fail(
+                            "invalid tool protocol after recovery"
+                        )
+                    await runtime.checkpoint(
```
Comment (rogercloud) [inline src/xagent/core/agent/pattern/react/react.py]: **Minor (duplication).** This invalid-tool-protocol failure block — the checkpoint call plus the `PatternResult(success=False, ...).to_dict()` construction — is duplicated near-verbatim between this `protocol_retry_performed` early-exit branch and the post-retry validation branch at lines 627-642. Only the preceding `answer_streamer.fail(...)` message differs ("after recovery" vs "after retry"). Consider extracting a shared helper, e.g. `_invalid_tool_protocol_result(iteration)`, taking the failure message as a parameter, to avoid the two copies drifting apart. Introduced by this PR (the base had a single block). Non-blocking.

File: src/xagent/core/model/chat/basic/deepseek_tool_protocol.py [hunk 6]
```diff
@@ -262,10 +332,106 @@ def _tool_call_violation(
...
+        elif char in "{[":
+            stack.append(char)
+        elif char in pairs:
+            if not stack or stack.pop() != pairs[char]:
+                return False
+
+    return string_quote is None and not escaped and not stack
+
+
+def _set_tool_call_arguments(tool_call: Any, arguments: str) -> bool:
```
Comment (rogercloud) [inline src/xagent/core/model/chat/basic/deepseek_tool_protocol.py:396]: **Minor (contract clarity).** `_set_tool_call_arguments` mutates the caller's `tool_call` in place (`function["arguments"] = arguments`, etc.), and `normalize_deepseek_response` returns the same object it was given. This in-place repair is not documented anywhere. Please add a one-line docstring note on `normalize_deepseek_response` (and/or here) stating that the input `tool_call`/response is repaired in place, so callers don't assume a fresh copy. Non-blocking.

File: src/xagent/core/agent/pattern/auto/auto.py [hunk 7]
```diff
@@ -776,7 +790,9 @@ async def _decide(
         )
         decision_tools = [self._decision_tool_schema()]
         retry_feedback: str | None = None
-        for attempt in range(MAX_DECISION_PARSE_ATTEMPTS):
+        attempt = 0
+        protocol_retries = 0
+        while attempt < MAX_DECISION_PARSE_ATTEMPTS:
```
Comment (rogercloud) [inline src/xagent/core/agent/pattern/auto/auto.py:796]: **Minor (clarity, not a bug).** `protocol_retries` is a single-use budget for the *entire* decision-parse while-loop, not reset per `attempt`. Edge case: if a protocol retry succeeds at attempt 0, a `RequiredToolCallError` retry advances to attempt 1, then a fresh unrelated protocol error arrives, the loop raises at the `protocol_retries >= 1` gate (~line 854) even though `MAX_DECISION_PARSE_ATTEMPTS` (2) still has room. This surfaces as a visible raised exception, not a silent wrong answer. A one-line comment here noting that this is a whole-loop budget, independent of `attempt`, would make the interaction clear. Non-blocking.

File: src/xagent/core/agent/pattern/react/react.py [hunk 8]
```diff
@@ -484,6 +487,52 @@ async def _run_tool_calling_loop(
...
+                        **llm_metadata,
+                        "phase": exc.code,
+                        "protocol_code": exc.code,
+                    },
+                )
+                try:
+                    (
+                        response,
+                        answer_streamer,
+                    ) = await self._retry_tool_protocol_response(
```
Comment (rogercloud) [inline src/xagent/core/agent/pattern/react/react.py:511]: **Minor / optional (pre-existing).** If the *retry* streaming call here itself raises a second consecutive `LLMToolProtocolError`, it propagates out of `_retry_tool_protocol_response` and out of `ReActPattern.run()` as an uncaught exception, rather than yielding the graceful `PatternResult(status="invalid_tool_protocol")` used when the retry's *response* (not a raised exception) is invalid. This gap is pre-existing — the base had identical propagation in the original retry path; this PR only added a parallel exception-raised path that mirrors it. It's also non-crashing, since `AgentRunner.run` and DAG step execution both wrap `pattern.run()` in `except Exception`. Optional follow-up: wrap the retry calls in `except LLMToolProtocolError` to normalize into the same `PatternResult` shape. Not required before merge.

**Review by rogercloud — APPROVED** [review]
## PR #943 — `fix: harden agent execution reliability`

### What this PR changes and why

This PR bundles a set of reliability fixes across the agent execution stack rather than one feature. The core change is structured handling of malformed/unavailable provider tool calls: a new `LLMToolProtocolError` (subclassing `LLMRetryableError`) drives **one bounded, schema-aware retry** in both the Auto and ReAct patterns, while `error.py`'s retry filter excludes two specific protocol codes so genuinely unfixable errors are not blindly replayed. Alongside this, DeepSeek's tool-protocol layer gains safe in-place repair of structurally-complete-but-malformed JSON tool arguments (preserving original diagnostics), DAG replanning/completion checks are made to respect valid dependencies and the user's authoritative requested scope, child Auto/DAG-step LLM errors are forwarded with correct execution metadata, file-delivery claims now require trusted tool-result provenance, checkpoint blob persistence is made idempotent under concurrent DAG writes (new `web/services/trace_message_storage.py`), and the image tool reports the actual model that served a request. 24 files, ~1750/-67, plus a new alembic merge migration and corresponding tests.

### Update summary (re-review)

All 4 prior Gemini bot findings are addressed as of HEAD `1ec05b6f`: **P1** (single-quote handling in JSON structural completeness) and **P2** (non-dict repair results) are fixed as literally suggested; **P3** (inline retry failure must stop execution) is shown to already be correct in the current revision — the bot appears to have reviewed a stale revision, and no change was needed; **P4** (object-shaped `_set_tool_call_arguments` fallback) is waived, with the author's design reasoning independently verified as factually correct. No new blocking issues surfaced in this round.

### Approach verdict — acceptable-with-reservations

The fixes are root-cause-oriented and land in the right architectural layers. `LLMToolProtocolError` subclassing `LLMRetryableError` plus the two-code exclusion in `error.py`'s retry filter is a clean way to stop blind-replay of unfixable protocol errors while keeping the retryable ones retryable. DeepSeek JSON repair lives in `deepseek_tool_protocol.py`, exactly where provider normalization already happens, and `trace_message_storage.py` hardens the existing blob-dedup service rather than duplicating persistence — its `on_conflict` `index_elements` were verified to match the DB unique constraints.

Two design-level reservations worth watching (neither blocking, and neither owned by a single line, so they are not filed as inline comments):

- **ReAct retry control flow is duplicated.** The "invalid tool protocol after one repair attempt" failure block is repeated near-verbatim between the new early-exit branch and the pre-existing post-retry branch (see finding #1). Extracting a shared helper would remove the drift risk.
- **Auto and ReAct implement the same "one bounded schema-aware retry" concept in structurally different ways** — Auto uses a while-loop with attempt/retry counters, ReAct uses inline `if`-blocks with flags. This is acceptable given the two patterns already differ structurally, but it is an inconsistency to keep an eye on as this retry logic evolves.

### Prior findings status

| ID | Severity | Area | Status | Justification |
|----|----------|------|--------|---------------|
| P1 | HIGH | `deepseek_tool_protocol.py` `_is_structurally_complete_json_object` | **FIXED** | Now tracks the active quote char (both `'` and `"`) with escape handling; regression test `test_deepseek_codec_repairs_brace_inside_single_quoted_string` passes on the exact adversarial case. |
| P2 | MEDIUM | `deepseek_tool_protocol.py` non-dict repair results | **FIXED** | Non-dict repairs get `repair_status="failed_non_dict"` and are treated as violations; empty `{}` deliberately kept valid (safe — this layer never enforces `required`). Both tests pass. |
| P3 | HIGH | `react.py:626` inline retry failure should stop | **FIXED** | Current code checkpoints and hard-`return`s `PatternResult(success=False, status="invalid_tool_protocol")`; test asserts the tool was called only once. Bot reviewed a stale revision. |
| P4 | MEDIUM | `deepseek_tool_protocol.py:411` object-shaped fallback | **WAIVED** | `.function` is the sole source of truth for object-shaped tool_calls; name validation returns before the fallback; downstream readers only read `tool_call.function.arguments`. Considered, correct design decision. |

### New findings (all minor — none blocking)

**#1 — ReAct duplicated failure block** — `src/xagent/core/agent/pattern/react/react.py:566-582` and `:627-642` — *minor (code quality).*
The invalid-tool-protocol failure block (checkpoint call + `PatternResult(success=False, ...).to_dict()`) is duplicated verbatim between the new `protocol_retry_performed` early-exit branch and the pre-existing post-retry validation branch; only the preceding `answer_streamer.fail(...)` message text differs ("after recovery" vs "after retry"). Introduced by this PR (base had one block). Extract a shared helper, e.g. `_invalid_tool_protocol_result(iteration)`, and pass the message in.

**#2 — In-place mutation is undocumented** — `src/xagent/core/model/chat/basic/deepseek_tool_protocol.py:394-410` — *minor (contract clarity).*
`_set_tool_call_arguments` mutates the caller's `tool_call` in place (`function["arguments"] = arguments`, etc.), and `normalize_deepseek_response` returns the same object it received. Nothing documents that repair mutates the input. Add a one-line docstring note on `normalize_deepseek_response` (and/or `_set_tool_call_arguments`) stating the input is repaired in place.

**#3 — Missing multi-chunk malformed streaming test** — `tests/core/model/chat/basic/test_deepseek_tool_protocol.py` — *minor (test coverage gap).*
No test exercises the accumulate-then-repair streaming path: an early chunk with incomplete JSON is buffered/skipped by `_arguments_are_ready_for_validation`, then a later chunk's now-complete-but-malformed accumulated arguments go through `repair_json_loads`. Existing tests cover either valid JSON split across chunks, or malformed JSON delivered whole in one chunk — never both together. This path is live in `adapt_deepseek_stream`/`_complete_streaming_tool_call_violation`. Add a test with malformed JSON split across 2+ streaming chunks.

**#4 — `protocol_retries` budget scope is non-obvious** — `src/xagent/core/agent/pattern/auto/auto.py:795` — *minor (clarity, not a bug).*
`protocol_retries` is a single-use budget for the *entire* decision-parse while-loop, not reset per `attempt`. If a protocol retry succeeds at attempt 0, a `RequiredToolCallError` retry advances to attempt 1, and a fresh unrelated protocol error then arrives, the loop raises at the `protocol_retries >= 1` gate (~line 854) even though the attempt budget (`MAX_DECISION_PARSE_ATTEMPTS = 2`) still has room. This surfaces as a visible raised exception, not a silent wrong answer — it just trims one rare corner case. Add a one-line comment at the `protocol_retries = 0` init noting it is a whole-loop budget, independent of `attempt`.

**#5 — Retry-call second protocol error propagates ungracefully** — `src/xagent/core/agent/pattern/react/react.py:511-524` and `:593-606` — *minor / optional (pre-existing).*
If the *retry* streaming call itself raises a second consecutive `LLMToolProtocolError`, it propagates out of `_retry_tool_protocol_response` and out of `ReActPattern.run()` as an uncaught exception, rather than yielding a graceful `PatternResult(status="invalid_tool_protocol")` (the sibling case, where the retry's *response* is invalid, is handled gracefully per #1). Note: this gap is **pre-existing** — the base commit had identical propagation in the original retry path; this PR only added a parallel exception-raised path mirroring it. It is also non-crashing: `AgentRunner.run` (`runner.py:155-178`) and DAG step execution (`dag.py:888`) both wrap `pattern.run()` in `except Exception`, so the process returns a generic failure instead of the specific `invalid_tool_protocol` status. Optional follow-up: wrap the retry calls in `except LLMToolProtocolError` to normalize into the same `PatternResult` shape. Not required before merge.

### Simplification opportunities

```
deepseek_tool_protocol.py
L394: yagni _set_tool_call_arguments handles 4 tool_call shapes; 2 are dead. Delete the dict-with-"args"-key fallback and the bare-dict-with-"arguments"-key-but-no-"function" fallback — traced through deepseek.py/openrouter.py → openai.py's uniform {"function": {...}} construction (streaming + non-streaming), no path produces them, no test covers them. KEEP the non-dict/object getattr/setattr branch (unreachable today but an intentional, verified forward-looking safety net — writing an outer attribute instead would silently leave function.arguments malformed while reporting success).
net: -10 lines possible
```

### Tests & coverage

335 tests pass at HEAD `1ec05b6f` (run during the design pass). The one coverage gap worth closing is finding #3 (multi-chunk + malformed streaming repair path). Optionally, a narrower note: `auto.py`'s generic `LLMToolProtocolError` fallback-message branch is reachable by real provider codes but not exercised by any `auto.py` test (`test_auto.py`'s protocol-error test is parametrized only over the 2 excluded codes) — a minor test-coverage gap, not dead code.

### Recommendation

**Approve with optional minor follow-ups.** All 4 prior findings are resolved, the fixes are correctly layered and root-cause-oriented, and there are zero blocking findings. The 5 new items are all minor (duplication cleanup, one docstring, a clarifying comment, a test-coverage gap, and a pre-existing non-crashing edge case) and can be handled as follow-ups rather than gating merge.

File: src/xagent/core/agent/pattern/react/react.py (hunk 5 above)
Comment (qinxuye) [reply src/xagent/core/agent/pattern/react/react.py]: Fixed in b091ddce. Both invalid-protocol exits now use a shared async helper that closes the answer stream, checkpoints, and constructs the failure result, while preserving the distinct stream failure messages.

File: src/xagent/core/model/chat/basic/deepseek_tool_protocol.py (hunk 6 above)
Comment (qinxuye) [reply src/xagent/core/model/chat/basic/deepseek_tool_protocol.py:396]: Fixed in b091ddce. normalize_deepseek_response now documents that successful malformed-argument repair mutates the supplied response in place.

File: src/xagent/core/agent/pattern/auto/auto.py (hunk 7 above)
Comment (qinxuye) [reply src/xagent/core/agent/pattern/auto/auto.py:796]: Fixed in b091ddce. Added a comment at protocol_retries initialization clarifying that it is a whole-loop budget independent of the parse-attempt budget.

File: src/xagent/core/agent/pattern/react/react.py (hunk 8 above)
Comment (qinxuye) [reply src/xagent/core/agent/pattern/react/react.py:511]: No code change: as noted, the second retry-call exception behavior is pre-existing, non-crashing, and optional. Keeping this PR scoped to the newly introduced reliability behavior; graceful normalization of that older edge case can be handled as a focused follow-up.
